app-backend/db/src/main/scala/MVTLayerDao.scala

```diff
@@ -54,40 +79,114 @@ object MVTLayerDao {
       z: Int,
       x: Int,
       y: Int
-  ): Query0[Array[Byte]] = {
-    fr"""WITH mvtgeom AS
-      (
+  ): Query0[LabelTileGeometry] = {
+    fr"""
         SELECT
-          ST_AsMVTGeom(
-            join_table_join.geometry,
-            ST_TileEnvelope(${z},${x},${y})
-          ) AS geom,
-          join_table_join.annotation_task_id,
+          annotation_labels.geometry,
+          ST_TileEnvelope(${z}, ${x}, ${y}) as envelope,
+          annotation_labels.annotation_task_id,
           annotation_label_classes.id as label_class_id,
           annotation_label_classes.name,
           annotation_label_classes.color_hex_code
         FROM
-          (annotation_labels join (select id, status from tasks) tasks on annotation_labels.annotation_task_id = tasks.id
+          (annotation_labels join tasks on annotation_labels.annotation_task_id = tasks.id
            JOIN annotation_labels_annotation_label_classes on
-           annotation_labels.id = annotation_labels_annotation_label_classes.annotation_label_id) join_table_join
-          JOIN annotation_label_classes on join_table_join.annotation_class_id = annotation_label_classes.id
+           annotation_labels.id = annotation_labels_annotation_label_classes.annotation_label_id)
+          JOIN annotation_label_classes on annotation_labels_annotation_label_classes.annotation_class_id = annotation_label_classes.id
         WHERE
           ST_Intersects(
-            join_table_join.geometry,
-            ST_TileEnvelope(${z},${x},${y})
+            annotation_labels.geometry,
+            ST_TileEnvelope(${z}, ${x}, ${y})
           )
-          AND join_table_join.annotation_project_id = ${annotationProjectId}
-          AND status <> 'SPLIT'
-      )
-    SELECT ST_AsMVT(mvtgeom.*) FROM mvtgeom;""".query[Array[Byte]]
+          AND annotation_labels.annotation_project_id = ${annotationProjectId}
+          AND tasks.status <> 'SPLIT'
+      """.query[LabelTileGeometry]
   }
 
+  /** We know the `foldMap` will produce an inhabited stream, since
+    * that's what the Monoid instance is for, so it's safe to compile it
+    * to a list and take the head.
+    */
+  @SuppressWarnings(Array("TraversableHead"))
   def getAnnotationProjectLabels(
       annotationProjectId: UUID,
       z: Int,
       x: Int,
       y: Int
-  ): ConnectionIO[Option[Array[Byte]]] =
-    getAnnotationProjectLabelsQ(annotationProjectId, z, x, y).option
+  ): ConnectionIO[Array[Byte]] = {
+    implicit val monoidStrictLayer = getMonoidStrictLayer(z, x, y)
+    getAnnotationProjectLabelsQ(annotationProjectId, z, x, y).stream
+      .foldMap({ labelTileGeom =>
```

Comment:
We can also do this in parallel with `parEvalMap(f).foldMonoid`, which I have on [some authority](https://gitter.im/functional-streams-for-scala/fs2?at=5dd6b373ac81632e65db2225) would never need to hold the whole stream in memory. I don't want to reach for that quite yet, since the function itself is pretty cheap, but it's hammer we can reach for if we need to